Sources/Smithy/Document+SmithyDocument.swift

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import struct Foundation.Data
+import struct Foundation.Date
+import class Foundation.DateFormatter
+
+extension Document: SmithyDocument {
+    func asBoolean() throws -> Bool {
+        guard case .boolean(let value) = self else {
+            throw SmithyDocumentError.typeMismatch("Expected boolean, got \(self)")
+        }
+        return value
+    }
+
+    func asString() throws -> String {
+        guard case .string(let value) = self else {
+            throw SmithyDocumentError.typeMismatch("Expected string, got \(self)")
+        }
+        return value
+    }
+
+    func asList() throws -> [Document] {
+        guard case .list(let value) = self else {
+            throw SmithyDocumentError.typeMismatch("Expected list, got \(self)")
+        }
+        return value
+    }
+
+    func asStringMap() throws -> [String: Document] {
+        guard case .map(let value) = self else {
+            throw SmithyDocumentError.typeMismatch("Expected map, got \(self)")
+        }
+        return value
+    }
+
+    func size() -> Int {
+        switch self {
+        case .list(let array): return array.count
+        case .map(let dict): return dict.count
+        default: return -1
+        }
+    }
+
+    func asByte() throws -> UInt8 {
+        let value = try asDouble()
```

Comment:
Why cast all these numeric types to Double before working with them?

Seems:
- if it's a Byte, just return it.
- if it's a number but not a Byte, see if you can create a byte from it without overflow, and return it.
- in any other case, throw.